File: apps/api/app/schemas/rto.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.enums import RTOStatus


class RTOUpdateRequest(BaseModel):
    status: RTOStatus | None = None
    notes: str | None = None


class RTOResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    shipment_id: uuid.UUID
    order_id: uuid.UUID
    courier_id: uuid.UUID | None
    reason: str | None
    normalized_reason: str | None
    external_reason: str | None
    status: RTOStatus
    initiated_at: datetime | None
    completed_at: datetime | None
    notes: str | None
    source_system: str | None
    created_at: datetime
    updated_at: datetime


class RTOListResponse(RTOResponse):
    """`RTOResponse` plus denormalized order/customer/product columns —
    see `NDRListResponse`'s docstring (app/schemas/ndr.py) for the
    identical convention this mirrors.
    """

    order_number: str | None = None
    customer_name: str | None = None
    customer_phone: str | None = None
    product: str | None = None
    order_amount: Decimal | None = None
    payment_type: str | None = None
    shipment_status: str | None = None

File: apps/web/types/rto.ts
// Mirrors app/schemas/rto.py

export type RTOStatus = "initiated" | "in_transit" | "received" | "cancelled"

export interface RTO {
  id: string
  shipment_id: string
  order_id: string
  courier_id: string | null
  reason: string | null
  normalized_reason: string | null
  external_reason: string | null
  status: RTOStatus
  initiated_at: string | null
  completed_at: string | null
  notes: string | null
  source_system: string | null
  created_at: string
  updated_at: string
  // Denormalized order/customer/product/shipment columns — present only
  // on rows from `GET /rto` (`RTOListResponse`).
  order_number?: string | null
  customer_name?: string | null
  customer_phone?: string | null
  product?: string | null
  order_amount?: string | null
  payment_type?: string | null
  shipment_status?: string | null
  awb?: string | null
  courier_name?: string | null
}

export interface RTOListFilters {
  q?: string
  status?: RTOStatus
  payment_type?: string
  courier_id?: string
  date_from?: string
  date_to?: string
}

export const RTO_STATUS_OPTIONS: { label: string; value: RTOStatus }[] = [
  { label: "Initiated", value: "initiated" },
  { label: "In transit", value: "in_transit" },
  { label: "Received", value: "received" },
  { label: "Cancelled", value: "cancelled" },
]
